base_utils.py

```
@@ -0,0 +1,267 @@
+                grouped['basic_info'].append(url)
+                grouped['about_and_contact_info'].append(url)  # Nhóm riêng biệt
+            elif raw_group == 'work_and_education':
+                grouped['work_and_education'].append(url)
+            elif raw_group == 'life_events':
+                grouped['life_events'].append(url)
+            elif raw_group == 'details':
+                grouped['details'].append(url)
+            # Bỏ qua các nhóm khác không cần thiết
+
+    # Đảm bảo luôn có tất cả các nhóm (dù là rỗng)
+    for group in expected_groups:
+        grouped.setdefault(group, [])
+
+    return dict(grouped)
+
+async def verify_client_api_key(x_api_key: str = Header(...)):
+    if x_api_key != os.getenv("API_KEY_VLM_SERVICE"):
+        raise HTTPException(status_code=403, detail="Forbidden: Invalid API Key")
+
+async def format_api_output_vlm(
+    dict_post_meta,
+    dict_person_meta,
+    dict_work2edu_meta,
+    dict_lifeEvent_meta,
+    dict_detail_meta,
+    dict_contact_meta  # <-- thêm contact
+):
+    # Lấy status và error
+    post_status = dict_post_meta.get("status", "fail")
+    profile_status = dict_person_meta.get("status", "fail")
+    workedu_status = dict_work2edu_meta.get("status", "fail")
+    event_status = dict_lifeEvent_meta.get("status", "fail")
+    detail_status = dict_detail_meta.get("status", "fail")
+    contact_status = dict_contact_meta.get("status", "fail")  # <-- thêm contact
+
+    # Xác định status tổng thể
+    status = (
+        "completed"
+        if all(s == "success" for s in [post_status, profile_status, workedu_status, event_status, detail_status, contact_status])
+        else "failed"
+    )
+
+    # error message chi tiết
+    error_message = {
+        "Profile": profile_status,
+        "Post": post_status,
+        "WorkEdu": workedu_status,
+        "Events": event_status,
+        "Detail": detail_status,
+        "Contact": contact_status,
```

Robustness: update_facebook_thirdparty swallows exceptions and logs them but does not propagate or return an error status. Consider returning a boolean or raising an exception so callers can react, or at least return response.status_code and text when non-200. Use logger.exception(...) in except to capture stack traces.